File: logica/generador.py
# -*- coding: utf-8 -*-
"""Generador de Fórmulas Bien Formadas aleatorias.

Su propósito es crear un **dataset** de fórmulas etiquetadas
(Tautología / Contradicción / Contingencia) a partir de tablas de verdad,
que servirá para entrenar el modelo de Machine Learning.

Estrategia para obtener clases balanceadas:
  - Si se quiere una tautología: se genera una F aleatoria y se usa ``F ∨ ¬F``
    (principio del tercero excluido: siempre es verdadera).
  - Si se quiere una contradicción: se usa ``F ∧ ¬F`` (siempre falsa).
  - Si se quiere una contingencia: se usa ``F`` tal cual (su tabla de verdad
    depende de la asignación).
"""

from __future__ import annotations

import random
from typing import List, Optional

from .parser import (
    AtomNode,
    NegNode,
    BinNode,
    a_cadena,
    atomos,
    render_es,
)
from .semantica import clasificar

#: Variables proposicionales disponibles para generar fórmulas.
ATOMOS_POOL = ["p", "q", "r", "s", "t"]

#: Banco de significados sencillos en español para las proposiciones.
DESCRIPCIONES_POOL = [
    "llueve", "hace frío", "hace calor", "nieva", "hay neblina",
    "estudio", "duermo", "como", "trabajo", "descanso",
    "corro", "bailo", "canto", "leo", "mira televisión", "viaja",
]

_OPERADORES_BINARIOS = ["∧", "∨", "→", "↔"]


def _fbf_aleatoria(vars_usadas: List[str], profundidad_restante: int):
    """Genera un AST aleatorio usando las variables dadas."""
    if profundidad_restante <= 0 or random.random() < 0.4:
        return AtomNode(random.choice(vars_usadas))

    modo = random.random()
    if modo < 0.25:
        return NegNode(_fbf_aleatoria(vars_usadas, profundidad_restante - 1))

    op = random.choice(_OPERADORES_BINARIOS)
    return BinNode(
        op,
        _fbf_aleatoria(vars_usadas, profundidad_restante - 1),
        _fbf_aleatoria(vars_usadas, profundidad_restante - 1),
    )


def _sobre_subconjunto():
    """Elige un subconjunto de 1 a 3 variables para 'sembrar' la fórmula."""
    n = random.randint(1, min(3, len(ATOMOS_POOL)))
    return random.sample(ATOMOS_POOL, n)


def _generar_formula(categoria: str) -> object:
    """Genera un AST cuya tabla de verdad corresponde (en su mayoría) a la clase."""
    vars_semilla = _sobre_subconjunto()
    base = _fbf_aleatoria(vars_semilla, profundidad_restante=random.randint(1, 3))

    if categoria == "tautologia":
        # F ∨ ¬F  es tautología garantizada.
        return BinNode("∨", base, NegNode(base))
    if categoria == "contradiccion":
        # F ∧ ¬F  es contradicción garantizada.
        return BinNode("∧", base, NegNode(base))
    # contingencia: fórmula al azar (se confirma luego con la tabla de verdad).
    return base


def proposicion_aleatoria(simbolos_permitidos: Optional[List[str]] = None) -> dict:
    """Genera una proposición al azar (simple o molecular) para practicar.

    Args:
        simbolos_permitidos: símbolos disponibles (p, q, r…). Por defecto usa
            todo :data:`ATOMOS_POOL`. Evita colisiones con tarjetas existentes.

    Returns:
        dict con ``formula`` (FBF canónica), ``descripciones`` (símbolo →
        significado) y ``lectura`` (la proposición en lenguaje natural).
    """
    pool = list(simbolos_permitidos if simbolos_permitidos else ATOMOS_POOL)
    n = min(random.randint(1, 3), len(pool))
    vars_usadas = random.sample(pool, n)
    descripciones = {v: random.choice(DESCRIPCIONES_POOL) for v in vars_usadas}

    if n == 1 and random.random() < 0.5:
        # Proposición simple (a veces negada).
        ast = AtomNode(vars_usadas[0])
        if random.random() < 0.3:
            ast = NegNode(ast)
    else:
        ast = _fbf_aleatoria(vars_usadas, random.randint(1, 2))

    # Ocasionalmente se niega el resultado completo…
    if random.random() < 0.25:
        ast = NegNode(ast)

    # …o se combina con un átomo nuevo para obtener una proposición molecular.
    libres = [s for s in pool if s not in vars_usadas]
    if libres and random.random() < 0.25:
        extra = random.choice(libres)
        descripciones[extra] = random.choice(DESCRIPCIONES_POOL)
        ast = BinNode(random.choice(_OPERADORES_BINARIOS), ast, AtomNode(extra))

    return {
        "formula": a_cadena(ast),
        "descripciones": descripciones,
        "lectura": render_es(ast, descripciones),
    }


def generar_instancia(categoria: str) -> dict:
    """Genera una instancia ``{formula, clase}`` con la clase ya verificada."""
    ast = _generar_formula(categoria)
    clas = clasificar(ast)
    texto = a_cadena(ast)
    indices = {
        "Tautología": "tautologia",
        "Contradicción": "contradiccion",
        "Contingencia": "contingencia",
    }
    # Si por azar la clase difiere de la pedida, la usamos igualmente porque
    # la etiqueta de verdad proviene siempre de la tabla de verdad.
    return {"formula": texto, "clase": clas["clase"], "target": indices.get(clas["clase"])}


def generar_dataset(n_objetivo: int = 1500, semilla: int = 42) -> List[dict]:
    """Genera un dataset balanceado de ``n_objetivo`` instancias etiquetadas.

    Returns:
        lista de dicts ``{"formula": str, "clase": str, "target": 0|1|2}``.
    """
    rng = random.Random(semilla)
    estado = rng.getstate()
    random.setstate(estado)          # semilla global para reproducibilidad

    cuotas = {
        "tautologia": n_objetivo // 3,
        "contradiccion": n_objetivo // 3,
        "contingencia": n_objetivo - 2 * (n_objetivo // 3),
    }
    acumulado = {"tautologia": [], "contradiccion": [], "contingencia": []}

    intentos = 0
    while len(acumulado["tautologia"]) < cuotas["tautologia"] \
            or len(acumulado["contradiccion"]) < cuotas["contradiccion"] \
            or len(acumulado["contingencia"]) < cuotas["contingencia"]:
        intentos += 1
        if intentos > 50_000:          # salvaguarda ante bucles infinitos.
            break

        categoria = random.choice(list(cuotas.keys()))
        instancia = generar_instancia(categoria)
        # La agrupamos por su clase real (la tabla de verdad manda).
        clave = {
            "Tautología": "tautologia",
            "Contradicción": "contradiccion",
            "Contingencia": "contingencia",
        }.get(instancia["clase"], "contingencia")
        try:
            if len(acumulado[clave]) < cuotas[clave]:
                acumulado[clave].append(instancia)
        except KeyError:
            acumulado["contingencia"].append(instancia)

    resultado: List[dict] = []
    # Mezclamos para que el dataset no quede agrupado por clase.
    for lista in acumulado.values():
        resultado.extend(lista)
    random.shuffle(resultado)
    return resultado


def resumen_dataset(datos: List[dict]) -> dict:
    """Cuenta cuántas instancias hay de cada clase (para mostrar en la GUI)."""
    conteo = {"Tautología": 0, "Contradicción": 0, "Contingencia": 0}
    for d in datos:
        if d["clase"] in conteo:
            conteo[d["clase"]] += 1
    return conteo


if __name__ == "__main__":
    # Prueba rápida de línea de comandos.
    import time

    inicio = time.time()
    data = generar_dataset(n_objetivo=300)
    print("Instancias generadas:", len(data))
    print("Distribución:", resumen_dataset(data))
    print("Ejemplos:")
    for d in data[:8]:
        print(f"  {d['formula']:<24} -> {d['clase']}")
    print(f"Tiempo: {time.time() - inicio:.2f}s")

File: web/assets/js/generador.js

"use strict";

(function () {
  var LP = window.LogicaParser;
  var S = window.Semantica;
  var G = {};

  G.ATOMOS_POOL = ["p", "q", "r", "s", "t"];
  G.DESCRIPCIONES_POOL = [
    "llueve", "hace frío", "hace calor", "nieva", "hay neblina",
    "estudio", "duermo", "como", "trabajo", "descanso",
    "corro", "bailo", "canto", "leo", "mira televisión", "viaja",
  ];

  var OPERADORES_BINARIOS = ["∧", "∨", "→", "↔"];


  G.alea = function (semilla) {
    var a = semilla >>> 0;
    return function () {
      a |= 0; a = (a + 0x6D2B79F5) | 0;
      var t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  };

  function fbfAleatoria(varsUsadas, profundidadRestante, rnd) {
    rnd = rnd || Math.random;
    if (profundidadRestante <= 0 || rnd() < 0.4) {
      return new LP.AtomNode(varsUsadas[Math.floor(rnd() * varsUsadas.length)]);
    }
    var modo = rnd();
    if (modo < 0.25) {
      return new LP.NegNode(fbfAleatoria(varsUsadas, profundidadRestante - 1, rnd));
    }
    var op = OPERADORES_BINARIOS[Math.floor(rnd() * OPERADORES_BINARIOS.length)];
    return new LP.BinNode(
      op,
      fbfAleatoria(varsUsadas, profundidadRestante - 1, rnd),
      fbfAleatoria(varsUsadas, profundidadRestante - 1, rnd)
    );
  }

  function sobreSubconjunto(rnd) {
    var n = 1 + Math.floor(rnd() * Math.min(3, G.ATOMOS_POOL.length));

    var pool = G.ATOMOS_POOL.slice();
    var res = [];
    for (var i = 0; i < n; i++) {
      var idx = Math.floor(rnd() * pool.length);
      res.push(pool.splice(idx, 1)[0]);
    }
    return res;
  }


  function generarForma(categoria, rnd) {
    var vars_semilla = sobreSubconjunto(rnd);
    var base = fbfAleatoria(vars_semilla, 1 + Math.floor(rnd() * 3), rnd);
    if (categoria === "tautologia") return new LP.BinNode("∨", base, new LP.NegNode(base));
    if (categoria === "contradiccion") return new LP.BinNode("∧", base, new LP.NegNode(base));
    return base;
  }


  G.proposicionAleatoria = function (simbolosPermitidos, rnd) {
    rnd = rnd || Math.random;
    var pool = simbolosPermitidos && simbolosPermitidos.length ? simbolosPermitidos.slice() : G.ATOMOS_POOL.slice();
    var n = Math.min(1 + Math.floor(rnd() * 3), pool.length);
    var varsUsadas = [];
    for (var i = 0; i < n; i++) {
      var idx = Math.floor(rnd() * pool.length);
      varsUsadas.push(pool.splice(idx, 1)[0]);
    }
    var descripciones = {};
    varsUsadas.forEach(function (v) {
      descripciones[v] = G.DESCRIPCIONES_POOL[Math.floor(rnd() * G.DESCRIPCIONES_POOL.length)];
    });

    var ast;
    if (n === 1 && rnd() < 0.5) {
      ast = new LP.AtomNode(varsUsadas[0]);
      if (rnd() < 0.3) ast = new LP.NegNode(ast);
    } else {
      ast = fbfAleatoria(varsUsadas, 1 + Math.floor(rnd() * 2), rnd);
    }
    if (rnd() < 0.25) ast = new LP.NegNode(ast);

    var libres = [];
    var usadas = {};
    varsUsadas.forEach(function (v) { usadas[v] = true; });
    libres = G.ATOMOS_POOL.filter(function (s) { return !usadas[s]; });
    if (libres.length && rnd() < 0.25) {
      var extra = libres[Math.floor(rnd() * libres.length)];
      descripciones[extra] = G.DESCRIPCIONES_POOL[Math.floor(rnd() * G.DESCRIPCIONES_POOL.length)];
      ast = new LP.BinNode(OPERADORES_BINARIOS[Math.floor(rnd() * OPERADORES_BINARIOS.length)], ast, new LP.AtomNode(extra));
    }

    return {
      formula: LP.aCadena(ast),
      descripciones: descripciones,
      lectura: LP.renderEs(ast, descripciones),
    };
  };


  G.generarInstancia = function (categoria, rnd) {
    var ast = generarForma(categoria, rnd);
    var cls = S.clasificar(ast);
    var indices = { Tautología: "tautologia", "Contradicción": "contradiccion", Contingencia: "contingencia" };
    var texto = LP.aCadena(ast);
    return { formula: texto, clase: cls.clase, target: indices[cls.clase] };
  };

  G.generarDataset = function (nObjetivo, semilla) {
    nObjetivo = nObjetivo || 1500;
    var rnd = G.alea(semilla == null ? 42 : semilla);
    var cuotas = {
      tautologia: Math.floor(nObjetivo / 3),
      contradiccion: Math.floor(nObjetivo / 3),
      contingencia: nObjetivo - 2 * Math.floor(nObjetivo / 3),
    };

    var acumulado = { tautologia: [], contradiccion: [], contingencia: [] };
    var claves = Object.keys(cuotas);
    var mapeo = { Tautología: "tautologia", "Contradicción": "contradiccion", Contingencia: "contingencia" };

    var intentos = 0;
    while (acumulado.tautologia.length < cuotas.tautologia ||
           acumulado.contradiccion.length < cuotas.contradiccion ||
           acumulado.contingencia.length < cuotas.contingencia) {
      intentos++;
      if (intentos > 50000) break;
      var categoria = claves[Math.floor(rnd() * claves.length)];
      var instancia = G.generarInstancia(categoria, rnd);
      var clave = mapeo[instancia.clase] || "contingencia";
      if (acumulado[clave].length < cuotas[clave]) acumulado[clave].push(instancia);
    }

    var resultado = [];
    Object.keys(acumulado).forEach(function (k) { resultado = resultado.concat(acumulado[k]); });

    for (var j = resultado.length - 1; j > 0; j--) {
      var r = Math.floor(rnd() * (j + 1));
      var tmp = resultado[j]; resultado[j] = resultado[r]; resultado[r] = tmp;
    }
    return resultado;
  };

  G.resumenDataset = function (datos) {
    var conteo = { Tautología: 0, "Contradicción": 0, Contingencia: 0 };
    datos.forEach(function (d) { if (conteo.hasOwnProperty(d.clase)) conteo[d.clase]++; });
    return conteo;
  };

  window.Generador = G;
})();